src/encoded/types/biosource.py

```diff
@@ -82,36 +82,114 @@ def biosource_name(self, request, biosource_type, individual=None,
             if tissue:
-                tissue_props = request.embed(tissue, '@@object')
-                if tissue_props.get('term_name') is not None:
-                    return tissue_props.get('term_name') + mod_str
+                tissue_props = get_item_if_you_can(request, tissue, 'ontology_terms')
+                if tissue_props:
+                    return tissue_props.get('preferred_name') + mod_str
                 else:
                     return biosource_type + mod_str
         elif biosource_type in cell_line_types:
             if cell_line:
-                cell_line_props = request.embed(cell_line, '@@object')
-                if cell_line_props.get('term_name') is not None:
-                    cell_line_name = cell_line_props.get('term_name')
+                cell_line_props = get_item_if_you_can(request, cell_line, 'ontology_terms')
+                if cell_line_props:
+                    cell_line_name = cell_line_props.get('preferred_name')
                     if cell_line_tier:
                         if cell_line_tier != 'Unclassified':
                             return cell_line_name + ' (' + cell_line_tier + ')' + mod_str
                     return cell_line_name + mod_str
             return biosource_type + mod_str
         elif biosource_type == "multicellular organism":
             if individual:
-                individual_props = request.embed(individual, '@@object')
-                organism = individual_props['organism']
-                organism_props = request.embed(organism, '@@object')
-                organism_name = organism_props['name']
+                organism_name = 'Unknown'
+                individual_props = get_item_if_you_can(request, individual, 'individuals')
+                if individual_props:
+                    organism = individual_props.get('organism')
+                    organism_props = get_item_if_you_can(request, organism, 'organisms')
+                    if organism_props:
+                        organism_name = organism_props['name']
                 return "whole " + organism_name + mod_str
         return biosource_type + mod_str
 
+    @calculated_property(schema={
+        "title": "Biosource categories",
+        "description": "Categories for the biosource.",
+        "type": "array",
+        "items": {
+            "title": "Category",
+            "type": "string"
+        }
+    })
+    def biosource_category(self, request, biosource_type, individual=None,
+                           cell_line=None, cell_line_tier=None, tissue=None,
+                           modifications=None):
+        oterms = self.registry['collections']['OntologyTerm']
+        tid2cat = {
+            'EFO:0003042': 'H1-hESC',
+            'EFO:0002784': 'GM12878',
+            '4DN:0000014': 'HFF (c6 or hTERT)',
+            'EFO:0009318': 'HFF (c6 or hTERT)',
+            '4DN:0000001': 'HFF (c6 or hTERT)',
+            '4DN:0000005': 'WTC-11',
```

Comment:
Would we want to also classify 4DN:0000267, 4DN:0000266, EFO:0009747 into WTC-11?

Reply:
Koray said that he was going to change the Biosources linked to the 4DN terms to WTC-11 and we should probably obsolete those 2 terms he created.  And I think we should probably give the EFO term the preferred name of WTC-11 and get rid of our current 4DN term for WTC-11.  We had created that term before we added preferred_name stuff to OntologyTerms I think or else I hadn't realized that GM25256 was the same as WTC-11.